WordPress/openverse-frontend
# Use talkback proxy to mock e2e API requests

## Fixes


Fixes #499 by @krysal

## Description


This PR adds a `talkback` proxy server that listens to the API requests during e2e testing. It saves the response data to `test/tapes` to be used on later requests.

This enables us to run e2e tests with real data, but without the internet access :)

I've also added a couple of pnpm scripts:
- `ci:e2e-test` runs a bash script that is to be run by the CI. It uses bash because our CI runs on Linux only. It runs the proxy server, builds the app for production, starts it and runs all the e2e tests.
- `update-tapes` removes all saved tapes, builds the server and runs the e2e tests, saving the responses to `test/tapes`.
There currently are no scripts that can be run with proxy server on Windows.
While testing on Windows I found that setting `--hostname=0` will not allow to build the app on Windows, so I changed it to cross-platform `0.0.0.0` value.

## Testing Instructions

1. Checkout the branch
2. `pnpm i` to install the new `talkback` dependency
3.  run `pnpm run ci:e2e-test`. You should see the proxy and build and test logs, and finally, a message that the tests were successful.
4. If you delete a couple of files in `test/tapes` and try to run again; some tests that use those tapes will fail.
5. Now, you can run `pnpm run update-tapes` and see that new tapes are added, and all the e2e tests pass.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
